src/workers/parser/mapBindings.js

```diff
@@ -0,0 +1,108 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+// @flow
+
+import { parseScript } from "./utils/ast";
+import { buildScopeList } from "./getScopes";
+import generate from "@babel/generator";
+import * as t from "@babel/types";
+import remove from "lodash/remove";
+
+import type { ColumnPosition } from "../../types";
+
+// translates new bindings `var a = 3` into `self.a = 3`
+// and existing bindings `var a = 3` into `a = 3` for re-assignments
+function globalizeDeclaration(node, bindings) {
+  return node.declarations.map(node => {
+    const identifier = bindings.includes(node.id.name)
+      ? node.id
+      : t.memberExpression(t.identifier("self"), node.id);
+
+    return t.expressionStatement(
+      t.assignmentExpression("=", identifier, node.init)
+    );
+  });
+}
+
+function globalizeAssignment(node, bindings) {
+  if (bindings.includes(node.left.name)) {
+    return node;
+  }
+
+  const identifier = t.memberExpression(t.identifier("self"), node.left);
+  return t.expressionStatement(
+    t.assignmentExpression("=", identifier, node.right)
+  );
+}
+
+function isTopLevel(ancestors) {
+  return ancestors.filter(ancestor => ancestor.key == "body").length == 1;
+}
+
+function replaceNode(ancestors, node) {
+  const ancestor = ancestors[ancestors.length - 1];
+
+  if (typeof ancestor.index === "number") {
+    ancestor.node[ancestor.key][ancestor.index] = node;
+  } else {
+    ancestor.node[ancestor.key] = node;
+  }
+}
+
+function replaceParent(ancestors, newNodes) {
+  const parent = ancestors[ancestors.length - 1];
+  const lastAncestor = ancestors[ancestors.length - 1];
```

Comment:
If this does stick around, `lastAncestor` can be removed here since it's the same as `parent`.